### PR TITLE
Add documentation for the aws-ecs-1 variant

### PR DESCRIPTION
**Pull request by samuelkarp**
**Issue number:**
https://github.com/bottlerocket-os/bottlerocket/issues/815

**Description of changes:**
* New `QUICKSTART-ECS.md` setup guide for ECS
* Updates to `README.md` to document the variant
* Updates to `BUILDING.md` to link to the new setup guide

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: QUICKSTART-ECS.md
Comment (srgothi92) [inline QUICKSTART-ECS.md]: I and sanika both landed on the same problem here, "" missing around CLUSTER_NAME and realized Instance was not getting attached to cluster. 
Just a Suggestion:
We can either add a note to validate the .toml file or make this as cluster = "CLUSTER_NAME". Even though example existed above it got neglected.

**Review by srgothi92 — COMMENTED** [review]
Additionally, Do you think we should add a line before `138` in your commit and `132` in old Readme.md section [Admin Container](https://github.com/bottlerocket-os/bottlerocket/blob/develop/README.md#admin-container) , conveying how to reach the `admin container` and `SSM session` is not the one directly connecting to `admin container`

File: QUICKSTART-ECS.md
Comment (srgothi92) [reply QUICKSTART-ECS.md]: Another Suggestion: A link to addtional userdata [settings](https://github.com/bottlerocket-os/bottlerocket#settings) from Readme.md file.

**Review by srgothi92 — COMMENTED** [review]
Do you think we should add supported feature list or TODO feature list somewhere ?

**Comment by samuelkarp** [comment]
> Additionally, Do you think we should add a line before 138 in your commit and 132 in old Readme.md section Admin Container , conveying how to reach the admin container and SSM session is not the one directly connecting to admin container

If you want to propose that as a separate PR, I think that could be reasonable.  We do already say at the top of that section: "This container has an SSH server that lets you log in as `ec2-user` using your EC2-registered SSH key."  But maybe we can make that more clear?

(I don't think that should be part of this PR as it's not an ECS-specific change.)

> Do you think we should add supported feature list or TODO feature list somewhere ?

I'm planning to update https://github.com/bottlerocket-os/bottlerocket/issues/815 with that information.

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done, thanks for the suggestions!

**Comment by srgothi92** [comment]
> Do you think we should add supported feature list or TODO feature list somewhere ?

 >> I’m planning to update #815 with that information.   

Shouldn’t that be somewhere public as well and linked to QuickStart-ECS doc, for anyone trying out ECS variant knows what features of ECS will work.

File: QUICKSTART-ECS.md
Comment (jahkeup) [inline QUICKSTART-ECS.md]: Should Fargate be mentioned here where folks reading how one uses Amazon ECS with *Bottlerocket* hosts? I'm worried this may cause some confusion. 

I'd be happy with just the pointers to the Amazon ECS docs taking up the torch on mentioning Fargate rather than adding them here. Is this out of consistency with the rest of the Amazon ECS docs?

Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Yep, this text was mostly lifted from the intro [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html).  I'm happy to change it to something else though.  (Or we could remove those sentences entirely?)

**Comment by samuelkarp** [comment]
> > I’m planning to update #815 with that information.
>
> Shouldn’t that be somewhere public as well and linked to QuickStart-ECS doc, for anyone trying out ECS variant knows what features of ECS will work.

Our GitHub issues are public.  I don't think it's the right choice to include this information in the quickstart/setup guide document as the primary audience of that is people trying out Bottlerocket for the first time rather than folks who are running their workloads on Bottlerocket.

We don't really have a section in our documentation today where we talk about what kinds of features do and don't work for our Kubernetes variants either.  I think #815 is a reasonable place to put the information for now, and we can potentially add a link to issues like @jahkeup suggested [here](https://github.com/bottlerocket-os/bottlerocket/pull/1052#issuecomment-676551795) once those are created.

File: QUICKSTART-ECS.md
Comment (jahkeup) [reply QUICKSTART-ECS.md]: This intro should stay scoped to the ECS bits that relate to Bottlerocket usages (at least in the QUICKSTART sense). The sprinkled pointers to the docs can help expand the subject and are better off introducing Fargate as there's more context to branch off that direction.

My vote is to remove the sentences and replace them with a shorter service introduction. Maybe something like:

```markdown
Amazon Elastic Container Service (Amazon ECS) is a highly scalable, fast, container management service that makes it easy to run, stop, and manage containers on a cluster.
Your containers are defined in a task definition which you use to run individual tasks or as a service.

This quickstart will walk through setting up an Amazon ECS cluster with Bottlerocket container instances (using EC2 Launch Type).
Check out the [Amazon ECS developer guide](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html) for more information on the supported Launch Types, additional details, and a broad overview of Amazon ECS.
```

<details>
<summary>rendered</summary>

Amazon Elastic Container Service (Amazon ECS) is a highly scalable, fast, container management service that makes it easy to run, stop, and manage containers on a cluster. Your containers are defined in a task definition which you use to run individual tasks or as a service.

This quickstart will walk through setting up an Amazon ECS cluster with Bottlerocket container instances (using EC2 Launch Type). Check out the [Amazon ECS developer guide](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html) for more information on the supported Launch Types, additional details, and a broad overview of Amazon ECS.

<details>

Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done, thanks for the suggestion!

File: QUICKSTART-ECS.md
Comment (srgothi92) [inline QUICKSTART-ECS.md]: *Check out the :)

Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Oops!

File: BUILDING.md
Comment (srgothi92) [inline BUILDING.md:7]: Shall we add info about how to build `ecs` image ? Basically, some info about setting `-e BUILDSYS_VARIANT=aws-ecs-1`

Comment (samuelkarp) [reply BUILDING.md:7]: Done

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
To get started using Bottlerocket with Kubernetes, please see [QUICKSTART-EKS](QUICKSTART-EKS.md).
```

The existing quickstart uses eksctl, so I wouldn't say it's general to k8s.

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
To get started using Bottlerocket with Amazon ECS, please see [QUICKSTART-ECS](QUICKSTART-ECS.md).
```

File: README.md
Comment (tjkirch) [inline README.md]: I'm wondering if we can combine these sections.  Something like "To get started using Bottlerocket with Kubernetes, please see bla.  For Amazon ECS, please see bla.  They describe: * how to set up a cluster, so your Bottlerocket instance can run containers; * how to launch bla"

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
The following settings are optional and allow you to configure how your instance joins an ECS cluster.
Since joining a cluster happens at startup, they need to be [specified in user data](#using-user-data).
```

**Comment by samuelkarp** [comment]
>  samuelkarp force-pushed the samuelkarp:ecs-documentation branch from b07a8ce to 1d32f2a now

This push adds documentation for the new `ecs.loglevel` setting added in https://github.com/bottlerocket-os/bottlerocket/pull/1062.

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
* `settings.ecs.logging-drivers`: The list of logging drivers available on the container instance.
```
Small tweak to hint that it's a list type.  I don't think we need an explicit example, though.

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
  The ECS agent running on a container instance must register the available logging drivers before containers placed on that instance can use log configuration options for those drivers in tasks.
```
It's quite a sentence.

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
  Supported values are `debug`, `info`, `warn`, `error`, and `crit`, and the default is `info`.
```

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md]:
```suggestion
[Amazon Elastic Container Service (Amazon ECS)](https://ecs.aws) is a highly scalable container management service that makes it easy to run, stop, and manage containers on a cluster.
```
I think "fast" is implied enough, but if you love it, no comma after :)

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md]:
```suggestion
Check out the [Amazon ECS developer guide](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html) for an overview of ECS.
```

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md]:
```suggestion
and that your AWS user has either the `AdministratorAccess` role or the permissions specified in the [Amazon ECS First Run Wizard Permissions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/security_iam_id-based-policy-examples.html#first-run-permissions) IAM policy example.
```

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md:13]: Also need aws-cli; copy from the existing quickstart?

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md:37]: I think when official ECS AMIs are published, we should move this into a separate doc, shared with the k8s quickstart.

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md]:
```suggestion
* BOTTLEROCKET_AMI_ID: the Amazon-provided AMI ID you found above, or the ID of an AMI you registered
```

File: QUICKSTART-ECS.md
Comment (tjkirch) [inline QUICKSTART-ECS.md]:
```suggestion
* INSTANCE_PROFILE_NAME: the IAM instance profile you created, e.g. `ecsInstanceRole`
```
This is going to be the case for almost everyone, so I don't think it hurts to reinforce this here

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: That sentence is taken verbatim from the first sentence [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html).  I think it's reasonable to keep them consistent, but if you think we should fix the comma in the developer guide as well we can do that too.

File: README.md
Comment (samuelkarp) [reply README.md]: I haven't updated this yet, but what do you think of something like this:

> Bottlerocket is best used with a container orchestrator. To get started with Kubernetes, please see [QUICKSTART-EKS]. To get started with Amazon ECS, please see [QUICKSTART-ECS].
>
> These guides describe:
> * how to set up a cluster with the orchestrator, so your Bottlerocket instance can run containers
> * how to launch a Bottlerocket instance in EC2

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

(And yeah...if you have suggestions on rephrasing it to make it less ungainly I'd welcome that too.)

Another possibility: The ECS agent running on a container instance must register the available logging drivers before tasks specifying the use of those drivers are eligible to be placed on that instance.

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: I made a slight tweak to your suggestion and linked to the [`AdministratorAccess` managed policy](https://console.aws.amazon.com/iam/home#policies/arn:aws:iam::aws:policy/AdministratorAccess).

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md:13]: Done, though I took out the reference to v1 and the request to use a recent one; any CLI newer than January 2015 is recent enough to work with ECS.

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done

File: BUILDING.md
Comment (jahkeup) [inline BUILDING.md:50]: `BUILDING.md` seems like a good place to hint at other controlling variables of interest, such as `BUILDSYS_ARCH`. Folks interested in building other variants may also want to build for another architecture.

File: README.md
Comment (webern) [inline README.md]: Wasn't one of your changes but seems better.

```suggestion
Our first supported variants, `aws-k8s-1.15`, `aws-k8s-1.16`, and `aws-k8s-1.17`, support EKS as described above.
```

**Review by webern — APPROVED** [review]
This part of QUICKSTART-EKS.md is my fault. It's even less true now than it was at the time:

```
Finally, you'll need [aws-cli v1](https://aws.amazon.com/cli/) set up to interact with AWS.
(You'll need a [recent v1 release](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html#install-tool-bundled) with EKS support.)
```

Please update that to leave out the version number and point to the latest release.

Edit: the links are correct as-is.

**Comment by samuelkarp** [comment]
> Please update that to leave out the version number and point to the latest release.

Looks like @srgothi92 is taking care of that in https://github.com/bottlerocket-os/bottlerocket/pull/1073.

File: BUILDING.md
Comment (samuelkarp) [reply BUILDING.md:50]: Agreed, but I'd prefer not to do that in the scope of this PR since generally documenting buildsys isn't quite the same thing as documenting the ECS variant...

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: BUILDING.md
Comment (jahkeup) [reply BUILDING.md:50]: I see as a new level of detail and, with it, there's an opportunity to add a brief note on a closely related flag. 

That said, I understand holding off for another PR - it's your call.

File: QUICKSTART-ECS.md
Comment (tjkirch) [reply QUICKSTART-ECS.md]: I think those official docs need to be more verbose than we do, in general, and we can pare down a bit overall.

File: README.md
Comment (tjkirch) [inline README.md]: nit: It won't let me leave a suggestion for whatever reason, but upon reading this and the ECS version in context, I think the "see the guide" note belongs above the "the following settings..." note; the note applies to the whole "Kubernetes/ECS settings" section, not just the first group of settings.  (And if we're changing it, I'd get rid of the `*`...)

File: README.md
Comment (tjkirch) [inline README.md]:
```suggestion
  The ECS agent running on a container instance must register available logging drivers before tasks that use those drivers are eligible to be placed on the instance.
```
Sort of a hybrid of our suggestions.

**Review by tjkirch — APPROVED** [review]

File: BUILDING.md
Comment (samuelkarp) [reply BUILDING.md:50]: I'm not going to include it in this PR.

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

File: README.md
Comment (samuelkarp) [reply README.md]: Done

**Review by bcressey — APPROVED** [review]

**Review by jahkeup — APPROVED** [review]

**Review by tjkirch — APPROVED** [review]

**Review by zmrow — APPROVED** [review]
🚴

File: QUICKSTART-ECS.md
Comment (rothgar) [inline QUICKSTART-ECS.md]: `.Vpcs[].VpcId` should be in quotes so the shell doesn't interpret it and give an error. Same with line 93

`'.Vpcs[].VpcId'`

File: QUICKSTART-ECS.md
Comment (rothgar) [inline QUICKSTART-ECS.md:76]: Instead of giving this warning throughout the doc would it be better to remove `--region` from the example commands and at the top of the doc say they should export `AWS_REGION`? It makes the experience better from a copy/paste workflow and if someone has a default region they don't need to worry about it.

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md]: What shell are you using that you're receiving errors with that?  The command as-written works fine for me with GNU bash 4.4.20.

File: QUICKSTART-ECS.md
Comment (samuelkarp) [reply QUICKSTART-ECS.md:76]: I'm planning to leave this as-is for now to keep it consistent with the EKS-focused guide.  We can make larger changes to both guides separately.

(Exported environment variables can also be more confusing to reproduce later; having an explicit command-line option helps for repeatability.)

File: QUICKSTART-ECS.md
Comment (tjkirch) [reply QUICKSTART-ECS.md]: zsh can fail in cases like this.

Comment (samuelkarp) [reply QUICKSTART-ECS.md]: Done

**Comment by samuelkarp** [comment]
[:arrow_up: This push](https://github.com/bottlerocket-os/bottlerocket/compare/e9bd57757a616d7c41ce9e0d7af3f9eca55cc9ee..cc687dcf6cb586e32e63235164db5437b976df1c) addresses @rothgar's comment about quoting the expression argument to `jq`.

**Pull request merged**
